examples/python/interface.py

```diff
@@ -0,0 +1,172 @@
+    # ---------------
+    # Set URL for testcase
+    url = 'http://127.0.0.1:5000'
+    # Set simulation parameters
+    length = config['length']
+    step = config['step']
+    # optional customzied kpi calculation
+    if "customized_kpi_config" in config:
+        customized_kpi_config = config['customized_kpi_config']
+    else:
+        customized_kpi_config = None 
+    if "prediction_config" in config:
+        prediction_config = config['prediction_config']
+    else:
+        prediction_config = None        
+    # ---------------
+
+    # TEST CONTROLLER IMPORT
+    # ----------------------
+    control = importlib.import_module(config["control_module"])
+    # ----------------------    
+
+    # GET TEST INFORMATION
+    # --------------------
+    print('\nTEST CASE INFORMATION\n---------------------')
+    # Test case name
+    name = requests.get('{0}/name'.format(url)).json()
+    print('Name:\t\t\t\t{0}'.format(name))
+    # Inputs available
+    inputs = requests.get('{0}/inputs'.format(url)).json()
+    print('Control Inputs:\t\t\t{0}'.format(inputs))
+    # Measurements available
+    measurements = requests.get('{0}/measurements'.format(url)).json()
+    print('Measurements:\t\t\t{0}'.format(measurements))
+    # Default simulation step
+    step_def = requests.get('{0}/step'.format(url)).json()
+    print('Default Simulation Step:\t{0}'.format(step_def))
+
+    # Define customized KPI if any
+    customizedkpis=[] # Initialize customzied kpi calculation list
+    customizedkpis_result={} # Initialize tracking of customized kpi calculation results
+    if customized_kpi_config is not None:
+        with open(customized_kpi_config) as f:
+                config=json.load(f,object_pairs_hook=collections.OrderedDict)
+        for key in config.keys():
+               customizedkpis.append(kpicalculation.cutomizedKPI(config[key]))
+               customizedkpis_result[kpicalculation.cutomizedKPI(config[key]).name]=[]
+    customizedkpis_result['time']=[]
+    # --------------------
+
+    # RUN TEST CASE
+    # -------------
+    start = time.time()
+    # Initialize test case
+    print('Initializing test case simulation.')
+    res = requests.put('{0}/initialize'.format(url), data={'start_time':0,'warmup_period':0}).json()
+    if res:
+        print('Successfully initialized the simulation')
+    print('\nRunning test case...')
+    # Set simulation step
+    res = requests.put('{0}/step'.format(url), data={'step':step})
+    # Initialize u
+    u = control.initialize()
+    # Store prediction if any
+    if prediction_config is not None:
+        predictions_store = pd.DataFrame(columns=prediction_config)
+    else:  
+        predictions_store = None    
+    # Simulation Loop
+    for i in range(int(length/step)):
+        # Advance simulation
+        y = requests.post('{0}/advance'.format(url), data=u).json()
+        # Compute next control signal                
+        if prediction_config is not None:

```

Comment:
I understand this is to accommodate the original implementation of testcase3.py.  However, I'm wondering if it is better to abstract the use of "predictions" to within each controller, rather than in this interface, so that this interface can be read/interpreted more easily.  Then, simply provide the ``/forecast`` result to the imported controller (e.g. as in line 124) and let the imported controller parse the forecast as needed.  Basically, move lines 120-123 to within the controller.  Then, maybe ``prediction_config`` just needs to be a True/False, which is easier to interpret when reading the interface code.  And I think it'd be ok to remove the storing and plotting (in testcase3.py) of forecasted set point values, otherwise, store within the controller and add a method called "get_setpoints" or something to be used later in the plotting.

Reply:
Should we make the "interface" a class and extend this class with addition attributes for testcase3?

Reply:
I'm not sure this is necessary, I'd be worried it may overcomplicate the readability of the code.  My suggestion relates to putting more of the details related to the parsing of the return ``forecast`` into the individual controller itself (since, abstractly, different controllers may utilize the predictions differently) rather than the generic interface, so as to make the reading of the generic interface code simpler.  That is, a reader sees whether a controller uses a forecast or not and how it may retrieve it.  If they want to know more about actually how this specific controller uses the forecast, they can go into the controller itself to view more.  But it is clear that the way the controller uses the forecast is specific to that controller, and not generic.

Reply:
As suggested, we put the forcast parsing into individual controller